simplejson/simplejson.go

```diff
@@ -0,0 +1,268 @@
+
+const QUERY_SEPARATOR = ";"
+const FIELDS_ITEMS_SEPARATOR = ","
+const DEFAULT_BACKEND = "tsdb"
+
+func SimpleJsonRequestFactory(method string, request []byte) (SimpleJsonRequestInterface, error) {
+	var retval SimpleJsonRequestInterface
+	switch method {
+	case "query":
+		retval = &SimpleJsonQueryRequest{Backend: DEFAULT_BACKEND}
+		break
+	case "search":
+		retval = &SimpleJsonSearchRequest{SimpleJsonQueryRequest{Backend: DEFAULT_BACKEND}}
+		break
+	default:
+		return nil, fmt.Errorf("Unknown method, %s", method)
+	}
+
+	if err := retval.ParseRequest(request); err != nil {
+		return nil, err
+	}
+	return retval, nil
+}
+
+func (req *SimpleJsonQueryRequest) GetReadRequest() *frames.ReadRequest {
+	return &frames.ReadRequest{Backend: req.Backend, Table: req.TableName, Columns: req.Fields, Start: req.From, End: req.To, Step: "60m",
+		Session: &pb.Session{Container: req.Container, User: req.Username, Password: req.Password},
+		Filter:  req.Filter}
+}
+
+func (req *SimpleJsonQueryRequest) formatKV(iter frames.FrameIterator) (interface{}, error) {
+	columns := []map[string]string{}
+	rows := [][]interface{}{}
+	for iter.Next() {
+		frame := iter.At()
+		iface, ok := frame.(pb.Framed)
+		if !ok {
+			return nil, errors.New("unknown frame type")
+		}
+		for _, column := range iface.Proto().Columns {
+			values, colType := req.extractColumnValuesTSDB(column)
+			columns = append(columns, map[string]string{"text": column.Name, "type": colType})
+			for rowIndex := 0; rowIndex < values.Len(); rowIndex++ {
+				if len(rows) <= rowIndex {
+					rows = append(rows, []interface{}{})
+				}
+				value := values.Index(rowIndex).Interface()
+				rows[rowIndex] = append(rows[rowIndex], value)
+			}
+		}
+	}
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	return []TableOutput{TableOutput{columns, rows, req.Type}}, nil
+}
+
+func (req *SimpleJsonQueryRequest) formatTSDB(iter frames.FrameIterator) (interface{}, error) {
+	retval := []TimeSeriesOutput{}
+	for iter.Next() {
+		frame := iter.At()
+		iface, ok := frame.(pb.Framed)
+		if !ok {
+			return nil, errors.New("unknown frame type")
+		}
+		target := req.getTargetTSDB(iface.Proto())
+		indices := frame.Indices()
+		for _, column := range iface.Proto().Columns {
+			if column.Name == "values" {
+				datapoints := [][]interface{}{}
+				values, _ := req.extractColumnValuesTSDB(column)
+				times, err := indices[0].Times()
+				if err != nil {
+					return nil, err
+				}
+				for j := 0; j < values.Len(); j++ {
+					datapoints = append(datapoints, []interface{}{values.Index(j).Interface(), req.formatTimeTSDB(times[j])})
+				}
+				retval = append(retval, TimeSeriesOutput{datapoints, target})
+			}
+		}
+	}
+	if iter.Err() != nil {
+		return nil, iter.Err()
+	}
+	return retval, nil
+}
+
+func (req *SimpleJsonQueryRequest) extractColumnValuesTSDB(column *pb.Column) (reflect.Value, string) {
+	colType := strings.ToLower(column.Dtype.String())
+	dataTypes := strings.Title(fmt.Sprintf("%ss", colType))
```

Comment:
I'd rather you'd do a `switch` on `column.DType`, this is relying on implementation details of column that might change.